Added solver for special case of 2nd order autonomous

hyadav2k opened the pull request:
This PR has added a solver for the special case of 2nd order autonomous of type `f''(x) = g(x) f'(x)^n`.

#### Brief description of what is added:
Mainly added the solver using the method specified in the Wikipedia article in which the first special case was already implemented and the second one is implemented in this PR. 

#### Release Notes
<!-- BEGIN RELEASE NOTES -->
* solvers
  * Added a new solver for the special case of 2nd order autonomous ode
<!-- END RELEASE NOTES -->

hyadav2k committed:
Added SecondAutonomousNth solver

hyadav2k committed:
Added solver for SecondAutonomousNth

hyadav2k committed:
removed unnecessary spaces

hyadav2k committed:
fixed docstring issues

oscarbenjamin commented:
If this is a generalisation of an existing solver is there any advantage in keeping the one that it generalises?

hyadav2k commented:
@oscarbenjamin, as of now, this PR doesn't handle the other case which is implemented in #20696, but I can compile both into a single class of  2nd order autonomous  which covers both the cases as mention in https://en.wikipedia.org/wiki/Autonomous_system_(mathematics) which are `f''(x) = g(f(x))` and `f''(x) = g(x) f'(x)**n`.
It just needs to add another `elif `condition for `n==0`.

Kindly provide your suggestion.
Thanks

oscarbenjamin commented:
There can possibly be advantages in having specific solver for particular special cases e.g. if it does not need to compute integrals or if it results in simpler integrals (the main limitation of these techniques is the need to compute two integrals).

Otherwise it would be good to aim for the most general method possible e.g.:
https://en.wikipedia.org/wiki/Autonomous_system_(mathematics)#Second_order

hyadav2k commented:
> @oscarbenjamin, as of now, this PR doesn't handle the other case which is implemented in #20696, but I can compile both into a single class of 2nd order autonomous which covers both the cases as mention in https://en.wikipedia.org/wiki/Autonomous_system_(mathematics) which are `f''(x) = g(f(x))` and `f''(x) = g(x) f'(x)**n`.
> It just needs to add another `elif `condition for `n==0`.
> 
> Kindly provide your suggestion.
> Thanks

Sorry it's my mistake both the cases can be done without adding any extra conditions, I missed `+/-` signs in the `else` condition.
Thank you for reminding

hyadav2k committed:
merged both cases and fixed docstirng

hyadav2k commented:
> Otherwise it would be good to aim for the most general method possible
>
Sure, working on it.

hyadav2k commented:
```
________________________________________________________________________________
________ sympy/solvers/ode/tests/test_single.py:test_2nd_autonomous_nth ________
Traceback (most recent call last):
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 389, in test_2nd_autonomous_nth
    _ode_solver_test(_get_examples_ode_sol_2nd_autonomous_nth)
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 173, in _ode_solver_test
    result = _test_particular_example(example['hint'], example, solver_flag=True)
  File "/home/runner/work/sympy/sympy/sympy/solvers/ode/tests/test_single.py", line 245, in _test_particular_example
    if sub_sol.has(Dummy):
AttributeError: 'list' object has no attribute 'has'
```
@oscarbenjamin, the failed test is showing this error. Can you please tell me how to debug a test file i.e. `test_single.py` in this case? 
Thanks.

oscarbenjamin commented on sympy/solvers/ode/tests/test_single.py:
```diff
@@ -1802,53 +1792,81 @@ def _get_examples_ode_sol_2nd_2F1_hypergeometric():
-            Eq(Integral(1/sqrt(C1 - 2*sinh(_u)), (_u, f(x))), C2 - x)
+            [Eq(sqrt(2)*Integral(1/sqrt(-3*_v**(S(4)/3) + 2*C1 - 4*log(_v)), (_v, f(x))), C2 + x),
+            Eq(sqrt(2)*Integral(1/sqrt(-3*_v**(S(4)/3) + 2*C1 - 4*log(_v)), (_v, f(x))), C2 - x)]
         ],
```
Here you've made this one a list of lists.

hyadav2k replied:
Thanks a lot for helping. I hope it passes all tests now.
Thanks.

hyadav2k committed:
fixes AttributeError

hyadav2k committed:
fixes AssertionError in test_ode.py

hyadav2k commented:
@oscarbenjamin, kindly review the PR.
Kindly suggest if some further betterment can be done.
Thank you for your time.

oscarbenjamin commented on sympy/solvers/ode/single.py:
```diff
@@ -791,66 +791,93 @@ def _get_general_solution(self, *, simplify: bool = True):
-    Gives solution for the autonomous second order nonlinear
-    differential equation of the form
+    The general second order autonomous differential equation of the form:
+    .. math :: x''(t) + f(x)*x'(t) + g(x) = 0
```
A general 2nd order autonomous ODE would be:
```
x''(t) = g(x'(t), x(t))
```

oscarbenjamin replied:
There should be a blank line before the equation

hyadav2k replied:
Sorry, I will correct it.

oscarbenjamin commented on sympy/solvers/ode/single.py:
```diff
@@ -791,66 +791,93 @@ def _get_general_solution(self, *, simplify: bool = True):
+
+    For proof of the results, kindly refer the below link:
+    https://en.wikipedia.org/wiki/Autonomous_system_(mathematics)
+    http://eqworld.ipmnet.ru/en/solutions/ode/ode0301.pdf
```
This should go in a references section:
https://docs.sympy.org/latest/documentation-style-guide.html

hyadav2k replied:
Sure, will do.

oscarbenjamin commented on sympy/solvers/ode/single.py:
```diff
@@ -791,66 +791,93 @@ def _get_general_solution(self, *, simplify: bool = True):
+    https://en.wikipedia.org/wiki/Autonomous_system_(mathematics)
+    http://eqworld.ipmnet.ru/en/solutions/ode/ode0301.pdf
+
+    Caution: If n is even except for 2, kindly set simplify = False in dsolve()
```
This needs an explanation

hyadav2k replied:
When `n` is even except for 2, it has a pair of solutions, so if we don't set `simplify` as `False`, it goes in the `_remove_redundant_solutions` which can't handle three equations in one soln i.e. in `soln1` in case of `f(x).diff(x, x) + f(x).diff(x)**4`, moreover,  `_remove_redundant_solutions` only handles simple cases.
Kindly give your opinion.
Thanks.

oscarbenjamin replied:
So if `n` is a symbol that is not known to be even or odd then I guess we can always return the pair and let `_remove_redundant_solutions` try to handle it. I think it's fine to sometimes return redundant solutions if necessary. Returning a Piecewise should be avoided unless it really is unavoidable.

oscarbenjamin commented on sympy/solvers/ode/single.py:
```diff
@@ -791,66 +791,93 @@ def _get_general_solution(self, *, simplify: bool = True):
-        lhs = Integral(1/sqrt(inside), (u, fx))
-        return [Eq(lhs, C2 + x), Eq(lhs, C2 - x)]
-
+        (C1, C2, C3) = self.ode_problem.get_numbered_constants(num=3)
```
Why does this need 3 constants? There should be only 2 constants for a 2nd order ODE.

hyadav2k replied:
My bad, I will correct it.

oscarbenjamin commented on sympy/solvers/ode/single.py:
```diff
@@ -791,66 +791,93 @@ def _get_general_solution(self, *, simplify: bool = True):
+            return [Eq(farg, C2 + x)]
+        else:
+            C3 = n - 2
+            if n%2 != 0:
```
What happens if `n` is a symbol?

hyadav2k replied:
Should a union of various piecewise equations be returned if `n `is a `symbol` instead of a `number`?

oscarbenjamin replied:
Is it possible to return a single expression that represents all cases?

hyadav2k replied:
In which case, when `n` is a `symbol` or for both cases when `n` is `symbol` and `number`?

oscarbenjamin replied:
Rather than checking whether or not `n` is a symbol or a number check the properties of `n` e.g. `if n.is_even`:
```python
In [1]: S(2).is_even
Out[1]: True

In [2]: S(3).is_even
Out[2]: False

In [3]: x = Symbol('x', even=True)

In [4]: x.is_even
Out[4]: True

In [5]: x.is_odd
Out[5]: False
```
That way `n` can be any arbitrary expression including something like `2*n + 1` where `n` is a an integer.

You need to be aware though that sometimes the result is `None` meaning unknown so there are always three cases to think about when checking these properties:
```python
In [6]: y = Symbol('y')

In [7]: print(y.is_even)
None
```
Does the method here require `n` to be an integer?

hyadav2k replied:
> Does the method here require `n` to be an integer?
>
No, `n` can be any arbitrary parameter.

hyadav2k commented:
@oscarbenjamin, I am trying something like:
``` py
c = C1*Integral(exp(Integral(a, v)), (v, fx))
d = Integral((C1 + (n - 2)*Integral(a, v))**(1/(n - 2)), (v, fx))
if n.is_even is None: 
    from sympy import Piecewise, Ne 
    return [Eq(fx, Piecewise((c - C2 - x, Eq(n, 2)), (d - C2 - x, Ne(n % 2, 0)), (d - C2 - x, Eq(n % 2, 0)), (d - C2 + x, Eq(n % 2, 0)) ))]
if n.is_even:
    if n == 2:
        return [Eq(c, C2 + x)]
    else:
        return [Eq(d, C2 + x), Eq(d, C2 - x)]
else:
    return [Eq(d, C2 + x)]
```
This implementation is giving correct results.
Should this be implemented?

oscarbenjamin commented:
Can you show an example equation and how the piecewise output looks?

hyadav2k commented:
@oscarbenjamin, in my terminal the symbols don't show correctly(some question mark comes instead of brackets and other operators), so after evaluating I stored the solution as `Eq(f(x), Piecewise((C1*exp(f(x)) + C2 - x, Eq(n, 2)), (C2 + x + Piecewise(((C1 + (n - 2)*f(x))**(1 + 1/(n - 2))/((1 + 1/(n - 2))*(n - 2)), Ne(1/(n - 2), -1)), (log(C1 + (n - 2)*f(x))/(n - 2), True)), True)))` and ran it on sympy live:

![t1](https://user-images.githubusercontent.com/45384915/105611448-6cd43680-5ddb-11eb-818b-f9e90763417c.PNG)

I ran it for `f(x).diff(x, x) + f(x).diff(x)**n`.

oscarbenjamin commented:
Are those solutions correct?

I tried with checkodesol:
```python
In [19]: eq = f(x).diff(x, x) + f(x).diff(x)**n

In [20]: sol = Eq(f(x), Piecewise((C1*exp(f(x)) + C2 - x, Eq(n, 2)), (C2 + x + Piecewise(((C1 + (n - 2)*f(x))**(1 + 1/(n - 2))/((1 + 1/(n - 2))*(n - 2)), Ne(1
    ...: /(n - 2), -1)), (log(C1 + (n - 2)*f(x))/(n - 2), True)), True)))

In [21]: checkodesol(eq.subs(n, 2), sol.subs(n, 2))
Out[21]: (False, 1)
```

hyadav2k commented:
@oscarbenjamin I am extremely sorry, it was a mix up of definitions of c and d in the terminal and sympy live,
I am not able to write piecewise for the case when `n % 2 == 0 `, how to return pair of solution,
`[ Piecewise( (c  - C2 - x, Eq(n, 2)), (d -  C2 - x, Ne(n % 2, 0)), ((d -  C2  + x, d - C2 -x), Eq(n % 2, 0)) ) ]` it gives TypeError:
```
TypeError                                 Traceback (most recent call last)
<ipython-input-2-75d60841ee73> in <module>
----> 1 sol=dsolve(eq, hint='2nd_autonomous_nth')

~\sympy\sympy\solvers\ode\ode.py in dsolve(eq, func, hint, simplify, ics, xi, eta, x0, n, **kwargs)
    661             # The key 'hint' stores the hint needed to be solved for.
    662             hint = hints['hint']
--> 663             return _helper_simplify(eq, hint, hints, simplify, ics=ics)
    664
    665 def _helper_simplify(eq, hint, match, simplify=True, ics=None, **kwargs):

~\sympy\sympy\solvers\ode\ode.py in _helper_simplify(eq, hint, match, simplify, ics, **kwargs)
    694             sols = solvefunc(eq, func, order, match)
    695         if iterable(sols):
--> 696             rv = [odesimp(eq, s, func, hint) for s in sols]
    697         else:
    698             rv =  odesimp(eq, sols, func, hint)

~\sympy\sympy\solvers\ode\ode.py in <listcomp>(.0)
    694             sols = solvefunc(eq, func, order, match)
    695         if iterable(sols):
--> 696             rv = [odesimp(eq, s, func, hint) for s in sols]
    697         else:
    698             rv =  odesimp(eq, sols, func, hint)

~\sympy\sympy\core\multidimensional.py in wrapper(*args, **kwargs)
    135                     result = apply_on_element(wrapper, args, kwargs, n)
    136                     return result
--> 137             return f(*args, **kwargs)
    138         return wrapper

~\sympy\sympy\solvers\ode\ode.py in odesimp(ode, eq, func, hint)
   2269         eq = simplify(eq)
   2270     if not isinstance(eq, Equality):
-> 2271         raise TypeError("eq should be an instance of Equality")
   2272
   2273     # Second, clean up the arbitrary constants.

TypeError: eq should be an instance of Equality
```

oscarbenjamin commented:
I think the error is because you are returning an expression rather than an equation.

hyadav2k commented:
If I try something like `[Eq(0, Piecewise( (c  - C2 - x, Eq(n, 2)), (d -  C2 - x, Ne(n % 2, 0)), ((d -  C2  + x, d - C2 -x), Eq(n % 2, 0)) )) ]`, then it gives `SymPyDeprecationWarning`:
```
AttributeError: 'Tuple' object has no attribute 'as_coeff_Mul'
```

oscarbenjamin commented:
It should be `Eq(f(x), ...)` not `Eq(0, ...)`

hyadav2k commented:
The problem is that, when `n` is a `symbol`, the integrals remain unevaluated, hence we can't equate it with `f(x) `as after solving the integral only, we will have `f(x) `in it, so I equated it to `0`. 

For `[Eq(fx, Piecewise( (c  - C2 - x, Eq(n, 2)), (d -  C2 - x, Ne(n % 2, 0)), ((d -  C2  + x, d - C2 -x), Eq(n % 2, 0)) )) ]`, 
it gives `AttributeError: 'Tuple' object has no attribute 'as_coeff_Mul'`. 

Again on setting `simplify=False`, no error arises and it returns:

`Eq(f(x), Piecewise((C1*exp(f(x)) - C2 - x, Eq(n, 2)), (-C2 - x + Piecewise(((C1 + (n - 2)*f(x))**(1 + 1/(n - 2))/((1 + 1/(n - 2))*(n - 2)), Ne(1/(n - 2), -1)), (log(C1 + (n - 2)*f(x))/(n - 2), True)), Ne(Mod(n, 2), 0)), ((-C2 + x + Piecewise(((C1 + (n - 2)*f(x))**(1 + 1/(n - 2))/((1 + 1/(n - 2))*(n - 2)), Ne(1/(n - 2), -1)), (log(C1 + (n - 2)*f(x))/(n - 2), True)), -C2 - x + Piecewise(((C1 + (n - 2)*f(x))**(1 + 1/(n - 2))/((1 + 1/(n - 2))*(n - 2)), Ne(1/(n - 2), -1)), (log(C1 + (n - 2)*f(x))/(n - 2), True))), True)))`

hyadav2k commented:
@oscarbenjamin, sorry for being inconsistent for several days, I had my internal examinations, now they are over.

The issue I am facing even if we return something like in `single.py`:
```
c = C1*Integral(exp(Integral(a, v)), (v, fx))
d = Integral((C1 + (n - 2)*Integral(a, v))**(1/(n - 2)), (v, fx))
if n.is_even is None:
    return [Eq(d, C2 + x), Eq(c, C2 + x), (Eq(d, C2 + x), Eq(d, C2 - x))]
elif n.is_odd:
    return [Eq(d, C2 + x)]
else:
    if n == 2:
        return [Eq(c, C2 + x)]
    else:
        return [Eq(d, C2 + x), Eq(d, C2 - x)]
```
 The problem arises in case when `n.is_even` returning 'None' or 'True' except for 'n=2' the tuple solution i.e. the conjugate solution `[Eq(d, C2 + x), Eq(d, C2 - x)]`, gets inside the `_remove_redundant_solutions` due to `simplify` being set `True`, thus it gives error as when `is_special_case_of` is called, solutions is in the form of list of solutions, which gives:
```
~\sympy\sympy\solvers\ode\ode.py in _is_special_case_of(soln1, soln2, eq, order, var)
   3889
   3890     # In case the solution is in implicit form we subtract the sides
-> 3891     soln1 = soln1.rhs - soln1.lhs
   3892     soln2 = soln2.rhs - soln2.lhs
   3893

AttributeError: 'list' object has no attribute 'rhs'
``` 
>
Even after fixing this by converting the list of list of solutions to list of solutions by:
```
temp=solns
for sol in solns:
    if isinstance(sol, list):
        temp.remove(sol)
        for solt in sol:
            temp.append(solt)       
solns=temp
```
It goes in some infinite loop and never returns any answer or some error. 
I don't know how to fix this since it does not return any error.
>
https://github.com/sympy/sympy/blob/9e8f62e059d83178c1d8a1e19acac5473bdbf1c1/sympy/solvers/ode/ode.py#L3850-L3858
>
Here if we use 'print(solns)' before entering the above loop, it gives the correct solution for all the cases of `n`.
For e.g. in `eq = f(x).diff(x, x) + f(x).diff(x)**4 `, this prints 
```
[Eq((C1 + 2*f(x))**(3/2)/3, C2 + x), [Eq(f(x), C1 + 6**(2/3)*(-C2**2 + 2*C2*x - x**2 + sqrt(C2**4 - 4*C2**3*x + 6*C2**2*x**2 - 4*C2*x**3 + x**4))**(1/3)/8 + 3*2**(2/3)*3**(1/6)*I*(-C2**2 + 2*C2*x - x**2 + sqrt(C2**4 - 4*C2**3*x + 6*C2**2*x**2 - 4*C2*x**3 + x**4))**(1/3)/8), Eq(f(x), C1 + 6**(2/3)*(-C2**2 + 2*C2*x - x**2 + sqrt(C2**4 - 4*C2**3*x + 6*C2**2*x**2 - 4*C2*x**3 + x**4))**(1/3)/8 - 3*2**(2/3)*3**(1/6)*I*(-C2**2 + 2*C2*x - x**2 + sqrt(C2**4 - 4*C2**3*x + 6*C2**2*x**2 - 4*C2*x**3 + x**4))**(1/3)/8), Eq(f(x), C1 - 6**(2/3)*(-C2**2 + 2*C2*x - x**2 + sqrt(C2**4 - 4*C2**3*x + 6*C2**2*x**2 - 4*C2*x**3 + x**4))**(1/3)/4)]]
```
To make it readable, I have indexed it below, solns[0] is the first line, and soln[1] is a list of three solutions as represented below:
```
[0]: Eq((C1 + 2*f(x))**(3/2)/3, C2 + x), 
[1][0]: Eq(f(x), C1 + 6**(2/3)*(-C2**2 + 2*C2*x - x**2 + sqrt(C2**4 - 4*C2**3*x + 6*C2**2*x**2 - 4*C2*x**3 + x**4))**(1/3)/8 + 3*2**(2/3)*3**(1/6)*I*(-C2**2 + 2*C2*x - x**2 + sqrt(C2**4 - 4*C2**3*x + 6*C2**2*x**2 - 4*C2*x**3 + x**4))**(1/3)/8), 
[1][1]: Eq(f(x), C1 + 6**(2/3)*(-C2**2 + 2*C2*x - x**2 + sqrt(C2**4 - 4*C2**3*x + 6*C2**2*x**2 - 4*C2*x**3 + x**4))**(1/3)/8 - 3*2**(2/3)*3**(1/6)*I*(-C2**2 + 2*C2*x - x**2 + sqrt(C2**4 - 4*C2**3*x + 6*C2**2*x**2 - 4*C2*x**3 + x**4))**(1/3)/8), 
[1][2]: Eq(f(x), C1 - 6**(2/3)*(-C2**2 + 2*C2*x - x**2 + sqrt(C2**4 - 4*C2**3*x + 6*C2**2*x**2 - 4*C2*x**3 + x**4))**(1/3)/4)
```
Here soln[1] are the explicit solutions i.e. they are in the form of `f(x) = `, whereas soln[0] is an implicit solution of one of the solutions, which it returns if we set `simplify=False `while calling `dsolve` and we get `[Eq((C1 + 2*f(x))**(3/2)/3, C2 + x), Eq((C1 + 2*f(x))**(3/2)/3, C2 - x)]` which is not simplified but no error raises here as it skips `_remove_redundant_solutions`.
>
Kindly guide me on how to move ahead.
Maybe the way I explained above would not be the best, kindly inquire if there is some point I made a little ambiguous or incomplete.

Thanks a lot for helping as always.

oscarbenjamin commented:
If the solution is coming out as a list of solutions when a single solution is expected then I think something has gone wrong somewhere earlier. It should always be known whether the solution is a list of solutions or not.

hyadav2k commented:
For `eq = f(x).diff(x, x) + f(x).diff(x)**3`, the value of `solns` is 
`[[Eq(f(x), -C1 - sqrt(C1**2 + C2 + 2*x)), Eq(f(x), -C1 + sqrt(C1**2 + C2 + 2*x))]]`.
>
For `eq = f(x).diff(x, x) + f(x).diff(x)**4`, the value of `solns` is  
> ```
> [Eq((C1 + 2*f(x))**(3/2)/3, C2 + x), [Eq(f(x), C1 + 6**(2/3)*(-C2**2 + 2*C2*x - x**2 + sqrt(C2**4 - 4*C2**3*x + 6*C2**2*x**2 - 4*C2*x**3 + x**4))**(1/3)/8 + 3*2**(2/3)*3**(1/6)*I*(-C2**2 + 2*C2*x - x**2 + sqrt(C2**4 - 4*C2**3*x + 6*C2**2*x**2 - 4*C2*x**3 + x**4))**(1/3)/8), Eq(f(x), C1 + 6**(2/3)*(-C2**2 + 2*C2*x - x**2 + sqrt(C2**4 - 4*C2**3*x + 6*C2**2*x**2 - 4*C2*x**3 + x**4))**(1/3)/8 - 3*2**(2/3)*3**(1/6)*I*(-C2**2 + 2*C2*x - x**2 + sqrt(C2**4 - 4*C2**3*x + 6*C2**2*x**2 - 4*C2*x**3 + x**4))**(1/3)/8), Eq(f(x), C1 - 6**(2/3)*(-C2**2 + 2*C2*x - x**2 + sqrt(C2**4 - 4*C2**3*x + 6*C2**2*x**2 - 4*C2*x**3 + x**4))**(1/3)/4)]]
> ```
> 
> To make it readable, I have indexed it below, solns[0] is the first line, and solns[1] is a list of three solutions as represented below:
> 
> ```
> [0]: Eq((C1 + 2*f(x))**(3/2)/3, C2 + x), 
> [1][0]: Eq(f(x), C1 + 6**(2/3)*(-C2**2 + 2*C2*x - x**2 + sqrt(C2**4 - 4*C2**3*x + 6*C2**2*x**2 - 4*C2*x**3 + x**4))**(1/3)/8 + 3*2**(2/3)*3**(1/6)*I*(-C2**2 + 2*C2*x - x**2 + sqrt(C2**4 - 4*C2**3*x + 6*C2**2*x**2 - 4*C2*x**3 + x**4))**(1/3)/8), 
> [1][1]: Eq(f(x), C1 + 6**(2/3)*(-C2**2 + 2*C2*x - x**2 + sqrt(C2**4 - 4*C2**3*x + 6*C2**2*x**2 - 4*C2*x**3 + x**4))**(1/3)/8 - 3*2**(2/3)*3**(1/6)*I*(-C2**2 + 2*C2*x - x**2 + sqrt(C2**4 - 4*C2**3*x + 6*C2**2*x**2 - 4*C2*x**3 + x**4))**(1/3)/8), 
> [1][2]: Eq(f(x), C1 - 6**(2/3)*(-C2**2 + 2*C2*x - x**2 + sqrt(C2**4 - 4*C2**3*x + 6*C2**2*x**2 - 4*C2*x**3 + x**4))**(1/3)/4)
> ```
So the problem is here as  suggested by you, the value of `solns` should have been  just `solns[1]` :
```
[[Eq(f(x), C1 + 6**(2/3)*(-C2**2 + 2*C2*x - x**2 + sqrt(C2**4 - 4*C2**3*x + 6*C2**2*x**2 - 4*C2*x**3 + x**4))**(1/3)/8 + 3*2**(2/3)*3**(1/6)*I*(-C2**2 + 2*C2*x - x**2 + sqrt(C2**4 - 4*C2**3*x + 6*C2**2*x**2 - 4*C2*x**3 + x**4))**(1/3)/8), Eq(f(x), C1 + 6**(2/3)*(-C2**2 + 2*C2*x - x**2 + sqrt(C2**4 - 4*C2**3*x + 6*C2**2*x**2 - 4*C2*x**3 + x**4))**(1/3)/8 - 3*2**(2/3)*3**(1/6)*I*(-C2**2 + 2*C2*x - x**2 + sqrt(C2**4 - 4*C2**3*x + 6*C2**2*x**2 - 4*C2*x**3 + x**4))**(1/3)/8), Eq(f(x), C1 - 6**(2/3)*(-C2**2 + 2*C2*x - x**2 + sqrt(C2**4 - 4*C2**3*x + 6*C2**2*x**2 - 4*C2*x**3 + x**4))**(1/3)/4)]]
```
Will try to find and fix it as to why `solns[0]` is included in `solns`.
Thanks.

hyadav2k commented:
@oscarbenjamin these changes solve our issue with setting unnecessary `simplify` condition in `dsolve`:

```
def _get_general_solution(self, *, simplify: bool = True):
        a, n = self.wilds_match()
        fx = self.ode_problem.func
        x = self.ode_problem.sym
        (C1, C2) = self.ode_problem.get_numbered_constants(num=2)
        v = Dummy('v')
        a = a.subs(fx, v)
        c = C1*Integral(exp(Integral(a, v)), (v, fx))
        d = Integral((C1 + (n - 2)*Integral(a, v))**(1/(n - 2)), (v, fx))
        if n.is_even is None:
            return [Eq(d, C2 + x)]
        elif n.is_odd:
            return [Eq(d, C2 + x)]
        else:
            if n == 0:
                d=d.subs(n,0)
                return [Eq(d, C2 + x), Eq(-d, C2 + x)]
            elif n == 2:
                return [Eq(c, C2 + x)]
            else:
                return [Eq(d, C2 + x), Eq(-d, C2 + x)]
```

Now we only need to set `simplify` as `False` on when n is undefined like in `eq=f(x).diff(x, x) + f(x).diff(x)**n`, here wolfram returns: https://www.wolframalpha.com/input/?i=f%27%27%28x%29+%3D+-f%27%28x%29%5En

![image](https://www4f.wolframalpha.com/Calculate/MSP/MSP13511290de0a446g0hd400005d5igd844ffg021e?MSPStoreType=image/gif&s=55)

And our answer as `[Eq(d, C2 + x)]` matches with it. So, for n being undefined we have to use: `sol=dsolve(eq, hint='2nd_autonomous_nth',simplify=False)`.

Kindly share your opinion.
Thanks.

naveensaigit commented:
I'm not sure if using `2nd_autonomous_nth` is good because `nth` usually refers to an nth order equation in all present hints. This might be confusing.

hyadav2k commented:
@naveensaigit, I added `nth` in the end and not just `2nd_autonomous` since this is not a general solver for any second-order autonomous equation, but only covers the special case of `x'' = f(x) x'**n`. As you mentioned, in the present hints, the `nth` is added at the start of the hints like in `nth_linear_constant_coeff_homogeneous`.
Kindly suggest, if some better hint name comes up.
Thanks.

hyadav2k committed:
Fixes unnecessary simplify condition

hyadav2k commented on sympy/solvers/ode/single.py:
```diff
@@ -825,23 +827,29 @@ class SecondAutonomousNth(SinglePatternODESolver):
+    Eq((C1 + 2*f(x))**(S(3)/2)/3, C2 + x)
+
+    >>> eq = f(x).diff(x, x) + f(x).diff(x)**n
+    >>> dsolve(eq, hint='2nd_autonomous_nth', simplify = False)
```
This can also be solved by not setting `simplify` to `False`.
```
>>> eq = f(x).diff(x, x) + f(x).diff(x)**n
>>> dsolve(eq, hint='2nd_autonomous_nth')
[Eq(f(x), Piecewise((C1/(n - 2) + (C2*n - C2 + n*x - x)**(n/(n - 1))/(n*(C2*n - C2 + n*x - x)**(2/(n - 1)) - 2*(C2*n - C2 + n*x - x)**(2/(n - 1))), Ne(1/(n - 2), -1)), (nan, True))), 
Eq(f(x), Piecewise((C1/(n - 2) + exp(C2*n + n*x)/(n*exp(2*C2 + 2*x) - 2*exp(2*C2 + 2*x)), Eq(1/(n - 2), -1)), (nan, True)))]
```

oscarbenjamin replied:
These are not correct for `n=1`.

hyadav2k commented on sympy/solvers/ode/single.py:
```diff
@@ -794,26 +794,28 @@ def _get_general_solution(self, *, simplify: bool = True):
+    This solver works for any arbitrary value of n defined above.
 
-    Caution: If n is even except for 2, kindly set simplify = False in dsolve()
+    Extras: For faster results, kindly set simplify = False in dsolve().
```
Now, we don't need to use `simplify` for any case as such. Though we can use it to get results faster as some solutions come in form of integrals that can't be solved further, and we can avoid `sympy` to not try simplifying it further, as it consumes a lot of time to find that it can't be simplified further by `sympy` like in `f(x).diff(x, 2) - exp(f(x)) + log(f(x))`.

hyadav2k committed:
fixes minor docs bug

hyadav2k committed:
fixes trailing whitespace

hyadav2k commented:
@oscarbenjamin, kindly review the PR.
Thanks.

oscarbenjamin commented on sympy/solvers/ode/single.py:
```diff
@@ -791,66 +791,101 @@ def _get_general_solution(self, *, simplify: bool = True):
+
+    >>> eq = f(x).diff(x, x) + f(x).diff(x)**4
+    >>> dsolve(eq, hint='2nd_autonomous_nth')
+    Eq((C1 + 2*f(x))**(3/2)/3, C2 + x)
```
There do not need to be so many examples in the docstring. Only one or possibly two examples are needed and they should be chosen carefully. Often the best example is just the general form of ODE that is solved by the method.

Examples that don't include `f(x)` are not interesting as they are more efficiently solved by reduction of order.

hyadav2k replied:
Sure, I removed all the examples and now it has just one example `f(x).diff(x, 2) + a*f(x).diff(x)**n`.
Will this work?
Thanks.

oscarbenjamin replied:
That example is better solved by reduction of order i.e. the substitution `g = f'`.

hyadav2k replied:
Will `f(x).diff(x, 2) + g(f(x))*f(x).diff(x)**n`?
It gives `Eq(Integral((C1 + (n - 2)*Integral(g(_v), _v))**(1/(n - 2)), (_v, f(x))), C2 + x)`.
We can't use `subs` since for different n values, there are different return types for equations.

oscarbenjamin replied:
Can we use a Piecewise to handle the different cases?

hyadav2k replied:
Something likes this works with the `subs` method:

```
def _get_general_solution(self, *, simplify: bool = True):
        a, n = self.wilds_match()
        fx = self.ode_problem.func
        x = self.ode_problem.sym
        (C1, C2) = self.ode_problem.get_numbered_constants(num=2)
        v = Dummy('v')
        a = a.subs(fx, v)
        c = C1*Integral(exp(Integral(a, v)), (v, fx))
        d = Integral((C1 + (n - 2)*Integral(a, v))**(1/(n - 2)), (v, fx))
        from sympy import Piecewise
        return Piecewise(((Eq(d, C2 + x), Eq(-d, C2 + x)), Eq(n,0)), (Eq(c, C2 + x), Eq(n, 2)), (Eq(d, C2 + x), True))
```

But this again arises the previous issue that if `n` is an undefined symbol, then set `simplify` as `False`, otherwise, it gives `ValueError` as `sympy` attempts to solve the integral, it imposes more Piecewise conditions on n in the Integral, hence it structs somehow in `odesimp` as: 

For `eq = f(x).diff(x, 2) + f(x).diff(x)**n` and `sol = dsolve(eq, hint='2nd_autonomous_nth')`:
```
<ipython-input-2-a6569a07ae34> in <module>
----> 1 sol = dsolve(eq, hint='2nd_autonomous_nth')

~\sympy\sympy\solvers\ode\ode.py in dsolve(eq, func, hint, simplify, ics, xi, eta, x0, n, **kwargs)
    661             # The key 'hint' stores the hint needed to be solved for.
    662             hint = hints['hint']
--> 663             return _helper_simplify(eq, hint, hints, simplify, ics=ics)
    664
    665 def _helper_simplify(eq, hint, match, simplify=True, ics=None, **kwargs):

~\sympy\sympy\solvers\ode\ode.py in _helper_simplify(eq, hint, match, simplify, ics, **kwargs)
    696             rv = [odesimp(eq, s, func, hint) for s in sols]
    697         else:
--> 698             rv =  odesimp(eq, sols, func, hint)
    699     else:
    700         # We still want to integrate (you can disable it separately with the hint)

~\sympy\sympy\core\multidimensional.py in wrapper(*args, **kwargs)
    135                     result = apply_on_element(wrapper, args, kwargs, n)
    136                     return result
--> 137             return f(*args, **kwargs)
    138         return wrapper

~\sympy\sympy\solvers\ode\ode.py in odesimp(ode, eq, func, hint)
   2269         eq = simplify(eq)
   2270     if not isinstance(eq, Equality):
-> 2271         raise TypeError("eq should be an instance of Equality")
   2272
   2273     # Second, clean up the arbitrary constants.

TypeError: eq should be an instance of Equality
```

For `eq = f(x).diff(x, 2) + f(x).diff(x)**n` and `sol = dsolve(eq, hint='2nd_autonomous_nth', simplify = False)`:
```
In [11]: print(sol)
Piecewise(((Eq(Piecewise(((C1 + (n - 2)*f(x))**(1 + 1/(n - 2))/((1 + 1/(n - 2))*(n - 2)), Ne(1/(n - 2), -1)), (log(C1 + (n - 2)*f(x))/(n - 2), True)), C2 + x), Eq(-Piecewise(((C1 + (n - 2)*f(x))**(1 + 1/(n - 2))/((1 + 1/(n - 2))*(n - 2)), Ne(1/(n - 2), -1)), (log(C1 + (n - 2)*f(x))/(n - 2), True)), C2 + x)), Eq(n, 0)), (Eq(C1*exp(f(x)), C2 + x), Eq(n, 2)), (Eq(Piecewise(((C1 + (n - 2)*f(x))**(1 + 1/(n - 2))/((1 + 1/(n - 2))*(n - 2)), Ne(1/(n - 2), -1)), (log(C1 + (n - 2)*f(x))/(n - 2), True)), C2 + x), True))
```

and

```
In [8]: sol.subs(n,2)
Out[8]:
    f(x)
C₁⋅ℯ     = C₂ + x

In [9]: sol.subs(n,4)
Out[9]:
             3/2
(C₁ + 2⋅f(x))
──────────────── = C₂ + x
       3

In [10]: sol.subs(n,1)
Out[10]: -log(C₁ - f(x)) = C₂ + x

```
So on summarising, if `n` is `undefined` then we need to set `simplify` as `False`, otherwise, we don't need to for any other case.

oscarbenjamin replied:
> otherwise, it gives `ValueError`

That just seems like a bug somehow that could be fixed. `get_general_solution` should return a list of `Eq`.

When I suggested a Piecewise I didn't mean a Piecewise that has `Eq` as its cases. I meant a list of Eqs where one of the sides of the Eqs is a Piecewise.

hyadav2k replied:
Will `[Eq(Piecewise((c, Eq(n, 2)), ((d, -d), Eq(n, 0)), (d, True)), C2+x)]` this work?
But this gives us `AttributeError: 'Tuple' object has no attribute 'as_coeff_Mul'`, how can we return two solutions for `n==0`.
Thanks.

oscarbenjamin replied:
Yes, it should be something like that. It should be possible to structure it so that you always return two Eqs but substituting some values of `n` makes them equal.

hyadav2k replied:
This should work `[Eq(Piecewise((d, Eq(n, 0)), (c, Eq(n, 2)), (d, True)), C2+x), Eq(Piecewise((-d, Eq(n, 0)), (c, Eq(n, 2)), (d, True)), C2+x)]`, but somehow it gives `AttributeError: 'list' object has no attribute 'rhs'` in `_remove_redundant_solutions`, though the idea is for `n==0` it will give pair of solutions and for other cases it will find both Equations to be same, thus, `_remove_redundant_solutions` should remove one of the solutions.

hyadav2k replied:
Returning this `[Eq(Piecewise((Piecewise((e,Eq(n,0)),(f,Eq(n, 1)),(d, True)), Ne(n, 2)), (nan, True)), C2+x), Eq(Piecewise((-e, Eq(n, 0)), (c, Eq(n, 2)), (nan, True)),C2+x)]` raises `RecursionError`, but no `AttributeError` as before.
I tried a lot of combinations to fix it, but nothing worked out for me, the issue is that `sympy` solves the integral further and imposes piecewise conditions on `n`, thus this makes `solns` to be` list of lists`, thus `_remove_redundant_solutions` doesn't work properly as it is supposed to have list of equations.

hyadav2k committed:
remove unnecessary examples and adds for genform

hyadav2k commented:
Update;
The only way we can make subs work is if we set `simpify` as `False` in `dsolve`, it will work as follows:
>
The below implementation will work for setting `simplify = False`:

```
def _get_general_solution(self, *, simplify: bool = True):
        a, n = self.wilds_match()
        fx = self.ode_problem.func
        x = self.ode_problem.sym
        (C1, C2) = self.ode_problem.get_numbered_constants(num=2)
        v = Dummy('v')
        a = a.subs(fx, v)
        c = C1*Integral(exp(Integral(a, v)), (v, fx))
        d = Integral((C1 + (n - 2)*Integral(a, v))**(1/(n - 2)), (v, fx))
        e = d.subs(n, 0)
        from sympy import Piecewise
        return [Eq(Piecewise((e, Eq(n, 0)), (c, Eq(n, 2)), (d, True)),  C2 + x ), Eq(Piecewise((-e, Eq(n, 0))),  C2 + x)]
```

Working:
```
In [1]: eq = f(x).diff(x,2) + f(x).diff(x)**n

In [2]: sol = dsolve(eq, hint='2nd_autonomous_nth',simplify=False)

In [3]: print(sol)
[Eq(Piecewise((-sqrt(C1 - 2*f(x)), Eq(n, 0)), (C1*exp(f(x)), Eq(n, 2)), (Piecewise(((C1 + (n - 2)*f(x))**(1 + 1/(n - 2))/((1 + 1/(n - 2))*(n - 2)), Ne(1/(n - 2), -1)), (log(C1 + (n - 2)*f(x))/(n - 2), True)), True)), C2 + x), Eq(Piecewise((sqrt(C1 - 2*f(x)), Eq(n, 0))), C2 + x)]

In [4]: print(sol[0].subs(n, 0))
Eq(-sqrt(C1 - 2*f(x)), C2 + x)

In [5]: print(sol[1].subs(n, 0))
Eq(sqrt(C1 - 2*f(x)), C2 + x)

In [6]: print(sol[0].subs(n, 2))
Eq(C1*exp(f(x)), C2 + x)

In [7]: print(sol[1].subs(n, 2))
False

In [8]: print(sol[0].subs(n, 4))
Eq((C1 + 2*f(x))**(3/2)/3, C2 + x)
```

The issue why the above implementation does not work without imposing any condition on `simplify` is that the term `(d, True)`, as when it goes for simplification further it gets broken into piecewise equations based on n-value and thus the return value becomes list of lists, which lead to `AttributeError: 'list' object has no attribute 'rhs'` in `_is_special_case_of`.

For `eq = f(x).diff(x,2) + f(x).diff(x)**n `, the return value is mention below, here we can see that the `(d, true)` term breaks into two piecewise equations for splitting at `Eq(1/(n - 2), -1)` as `simplify` further solves the integral. So, this list of lists raises the `AttributeError`.  
This value is obtained by printing `solns` in `_remove_redundant_solutions`. 
```
[[Eq(f(x), Piecewise((C1/(n - 2) + (C2*n - C2 + n*x - x)**(n/(n - 1))/(n*(C2*n - C2 + n*x - x)**(2/(n - 1)) - 2*(C2*n - C2 + n*x - x)**(2/(n - 1))), (n > 2) | (n < 0) | ((n > 0) & (n < 1)) | ((n > 1) & (n < 2))), (nan, True))), Eq(f(x), Piecewise((C1/(n - 2) + exp(C2*n + n*x)/(n*exp(2*C2 + 2*x) - 2*exp(2*C2 + 2*x)), Eq(1/(n - 2), -1)), (nan, True))), Eq(f(x), Piecewise((log(C1*C2 + C1*x), Eq(n, 2)), (nan, True)))], Eq(f(x), Piecewise((C1 - C2**2/2 - C2*x - x**2/2, Eq(n, 0)), (nan, True)))]
```
>
>
@oscarbenjamin, kindly show the way ahead. Should we be returning only one set of equations with `Piecewise` in it by setting `simplify = False` OR we should return separately without imposing any condition on `simplify` but can't use `subs` on `sol` for `n` if `n.is_symbol` is `True`.

Kindly share your opinion.
Thanks.

oscarbenjamin commented:
> The issue why the above implementation does not work without imposing any condition on `simplify` is that the term `(d, True)`, as when it goes for simplification further it gets broken into piecewise equations based on n-value and thus the return value becomes list of lists, which lead to `AttributeError: 'list' object has no attribute 'rhs'` in `_is_special_case_of`.

I don't understand exactly what this means but it sounds like something that should be fixed rather than worked around.

hyadav2k committed:
fixes simplify issues

hyadav2k committed:
removes redundant code

oscarbenjamin commented on sympy/solvers/ode/ode.py:
```diff
@@ -694,6 +694,10 @@ def _helper_simplify(eq, hint, match, simplify=True, ics=None, **kwargs):
             sols = solvefunc(eq, func, order, match)
         if iterable(sols):
+            for i, arg in enumerate(sols):
+                if isinstance(arg, BooleanFalse):
```
I don't think this should be handled here. Where does the BooleanFalse come from?

hyadav2k replied:
I am extremely sorry, I will be adding a hint condition so it doesn't bother other cases.
Thanks.

oscarbenjamin replied:
It should still be handled somewhere else. The BooleanFalse should not be there and we should not add code here to workaround that problem.

oscarbenjamin commented on sympy/solvers/ode/ode.py:
```diff
@@ -711,6 +715,18 @@ def _helper_simplify(eq, hint, match, simplify=True, ics=None, **kwargs):
+            for i, arg in enumerate(rv):
+                if i >= len_rv:
+                    break
+                if isinstance(arg, list):
```
This also should not be handled here. Where does the list come from?

This list should not have been inserted in place of an equation in the first place. We need to find where it comes from and stop it from being there.

hyadav2k replied:
I am sorry, but I didn't get which list you are talking about.
Here I am making these changes in the `ode` module so that we can return a list of solutions and as returning two solutions for the case of `n == 0` makes it extremely tough to find a way other than the way I did it( at least for me).
Thanks.

oscarbenjamin replied:
If this code here is needed then it means that what is actually being returned (from somewhere else) is a list of lists. It should just be a list of Eq. We shouldn't try to fix that here. It should be fixed wherever a list is being used in place of an Eq.

hyadav2k committed:
adds hint condition

hyadav2k commented:
This is how few examples will look like:
```
In [1]: eq = f(x).diff(x, 2) + f(x).diff(x)**n

In [2]: sol = dsolve(eq, hint = '2nd_autonomous_nth')

In [3]: print(sol)
[Eq(f(x), Piecewise((C1 - C2**2/2 + C2*x - x**2/2, Eq(n, 0)), (nan, True))), Eq(f(x), Piecewise((C1/(n - 2) + (C2*n - C2 + n*x - x)**(n/(n - 1))/(n*(C2*n - C2 + n*x - x)**(2/(n - 1)) - 2*(C2*n - C2 + n*x - x)**(2/(n - 1))), (n > 2) | (n < 0) | ((n > 0) & (n < 1)) | ((n > 1) & (n < 2))), (nan, True))), Eq(f(x), Piecewise((C1 - C2**2/2 - C2*x - x**2/2, Eq(n, 0)), (nan, True))), Eq(f(x), Piecewise((C1 + C2*exp(-x), Eq(1/(n - 2), -1)), (nan, True))), Eq(f(x), Piecewise((log(C1*C2 + C1*x), Eq(n, 2)), (nan, True)))]

In [4]: sol[0].subs(n, 0)
Out[4]:
              2           2
            C₂           x
f(x) = C₁ - ─── + C₂⋅x - ──
             2           2

In [5]: sol[1].subs(n, 0)
Out[5]: False

In [6]: sol[2].subs(n, 0)
Out[6]:
              2           2
            C₂           x
f(x) = C₁ - ─── - C₂⋅x - ──
             2           2

In [7]: sol[4].subs(n, 2)
Out[7]: f(x) = log(C₁⋅C₂ + C₁⋅x)

```
Here, in `eq =  f(x).diff(x, 2) + f(x).diff(x)**n`, it returns a list of solutions for different cases of `n`, now , we can check as above for solutions.
 
```
In [8]: eq = f(x).diff(x, x) + f(x).diff(x)**2

In [9]: sol = dsolve(eq, hint = '2nd_autonomous_nth')

In [10]: print(sol)
Eq(f(x), log(C1*(C2 + x)))

```

hyadav2k committed:
fixes simplify in odesimp

hyadav2k commented:
@oscarbenjamin, Kindly review the latest commit.
I added the code in `odesimp`, since it was only the one creating all the issues, so did a specific implementation for `2nd_autonomous_nth`.
Thanks.

oscarbenjamin commented:
The `odesimp` function is already messy and this makes it worse. We should be trying to eliminate special case code like this rather than adding more of it.

I don't understand exactly what problem you are trying to fix by changing `odesimp` or anything else here but I don't think that this is the right approach. If there's something wrong with the way odesimp works for this particular solver then there's a good chance that it is also problematic for other solvers. It should be fixed in general not just for some particular cases.

hyadav2k commented:
It fixes the `simplify` conditions so that we can make the solver work without providing any conditions on `simplify`. Secondly, when `n` is not specified in the ODE like `eq = f(x).diff(x, 2) + f(x).diff(x)**n` here, we return a list of solutions for any given `n` in the ODE, where `subs` can be used later on the returned list to get the solutions for any n-value.

In any other solvers, we don't return a pair of Piecewise Equations, like in this case, hence, `odesimp` works fine for them. 
For this solver, the issue with `odesimp` is that it returns a list of lists of solutions where a list of the solution was supposed to return, moreover, it doesn't handle cases like if some Piecewise returns False, thus in the code I fixed them.
 `Odesimp` doesn't allow to return two solutions for a single Piecewise condition, and for n being a symbol, like `Eq(Piecewise((e,-e), Eq(n, 0)))`.
It also simplifies for getting `f(x)` in `lhs` as `Eq(f(x), rhs)` without considering there could be two cases if power of f(x) is even(like f(x)**2 = C1, here it has 2 solutions, but when it solves for general n, it just simplifies for Eq(f(x), rhs), without considering it for n values).

I tried to make it work in general, but the way `odesimp` works, I strongly believe changing in `odesimp` will definitely raise new errors.
Kindly share your opinion.
Thanks.

hyadav2k commented:
There are few things we can fix in `odesimp`:

1. It should handle a case when the `eq` is `False`, suppose we are returning many Piecewise equations then there is a chance that one of the equations could return False if the condition doesn't satisfy, like in our case of `n != 0`. This is easy to fix, we can just add a usual if condition as:
```
if isinstance(eq, BooleanFalse):
            return None
```
I have already manipulated the `rv` list so that it ignores the None return type and doesn't add them in the list `rv`, in
https://github.com/sympy/sympy/blob/9e8f62e059d83178c1d8a1e19acac5473bdbf1c1/sympy/solvers/ode/ode.py#L697
To
```
rv = []
for s in sols:
    t = odesimp(eq, s, func, hint)
    if t is not None:
        rv.append(t)
```
This passes all the tests.

2. In the case when a Piecewise equation is given, it returns as a list of Piecewise equations, rather than just returning them as a single equation containing Piecewise solutions. For e.g. for `eq = f(x).diff(x, 2) + f(x).diff(x)**n`, we can see when a Piecewise equation is given to it as

   `Eq(Piecewise((-log(-C1 + f(x)), Eq(1/(n - 2), -1)), (C1*exp(f(x)), Eq(n, 2)), (sqrt(C1 - 2*f(x)), Eq(n, 0)), (Piecewise(((C1 + (n - 2)*f(x))**(1 + 1/(n - 2))/((1 + 1/(n - 2))*(n - 2)), Ne(1/(n - 2), -1)), (log(C1 + (n - 2)*f(x))/(n - 2), True)), True)), C2 + x)`,

   Here  `odesimp` returns us:

   `[Eq(f(x), Piecewise((-C1/(n - 2) + (C2*n - C2 + n*x - x)**(n/(n - 1))/(n*(C2*n - C2 + n*x - x)**(2/(n - 1)) - 2*(C2*n - C2 + n*x - x)**(2/(n - 1))), (n > 2) | (n < 0) | ((n > 0) & (n < 1)) | ((n > 1) & (n < 2))), (nan, True))), Eq(f(x), Piecewise((C1/2 - C2**2/2 - C2*x - x**2/2, Eq(n, 0)), (nan, True))), Eq(f(x), Piecewise((C1 + exp(-C2 - x), Eq(1/(n - 2), -1)), (nan, True))), Eq(f(x), Piecewise((log(C2/C1 + x/C1), Eq(n, 2)), (nan, True)))]`

   which is a list of Piecewise equations. Moreover another issue is that it solves for the Piecewise `Piecewise(((C1 + (n - 2)*f(x))**(1 + 1/(n - 2))/((1 + 1/(n - 2))*(n - 2)), Ne(1/(n - 2), -1)), (log(C1 + (n - 2)*f(x))/(n - 2), True))`, it solves for f(x), without consider what if the exponent of `f(x)` is even and may give two solutions instead of `n`.
   Occurs at:
   https://github.com/sympy/sympy/blob/9e8f62e059d83178c1d8a1e19acac5473bdbf1c1/sympy/solvers/ode/ode.py#L2330

These two issues are major in `odesimp`, which can be fixed for general cases, still, there will be one case that can't be fixed in general. It doesn't allow us to return two solutions for satisfying a single Piecewise, like in our case for `n==0`, we need to return two solutions, but `odesimp` doesn't allow them to return as Tuple or List.

I will try to reduce it further and fix it in general, but there will be this 2nd issue when n is a symbol, we need to do the special implementation for this case so that we can use the `subs` method for the solutions returned by it.
Thanks.

hyadav2k committed:
removes avoidable odesimp additions

hyadav2k commented:
@oscarbenjamin, kindly review the latest commit. I have removed other cases from the hint condition and simplified it. This hint condition is an absolute necessity because we don't want to `solve `to run in the case when `n `is a `Symbol`. The problem is that `solve `, solves for `f(x)` and ignores the case of multiple solutions as mentioned above. Moreover, we further don't want to get a list of Equations with Piecewise in them for a single equation. 

https://github.com/sympy/sympy/blob/9e8f62e059d83178c1d8a1e19acac5473bdbf1c1/sympy/solvers/ode/ode.py#L2330

I tried really hard to remove the additional hint condition in `odesimp`, but the only thing we want from the `odesimp` is: 
https://github.com/sympy/sympy/blob/9e8f62e059d83178c1d8a1e19acac5473bdbf1c1/sympy/solvers/ode/ode.py#L2264

Kindly share your opinion.
Thanks.

oscarbenjamin commented:
> These two issues are major in `odesimp`, which can be fixed for general cases, still, there will be one case that can't be fixed in general. It doesn't allow us to return two solutions for satisfying a single Piecewise, like in our case for `n==0`, we need to return two solutions, but `odesimp` doesn't allow them to return as Tuple or List.

I don't follow this. If there were two solutions then wouldn't they be passed to `odesimp` separately?

hyadav2k commented:
> I don't follow this. If there were two solutions then wouldn't they be passed to `odesimp` separately?
>
I am trying to say that it doesn't allow something like return `[ Eq( Piecewise( (e, -e), Eq(n, 0) ), C2 + x), Eq(Piecewise....) ]`, as passing them as `Tuple` raises error which is obvious since we didn't impose any check for this case, so we need to return them in different equations as `[ Eq ( Piecewise( (e, Eq(n, 0)) ), C2 + x), Eq ( Piecewise( (-e, Eq(n, 0)) ), C2 + x)]`.

Thanks.

hyadav2k commented on sympy/solvers/ode/ode.py:
```diff
@@ -2259,6 +2262,23 @@ def odesimp(ode, eq, func, hint):
     """
+    if isinstance(eq, BooleanFalse):
+        return None
+    if hint == '2nd_autonomous_nth':
```
This `if` condition is only implemented for the case when `n` is a symbol in the ode passed. Otherwise, if `n` is an arbitrary number, this condition is not useful to us. This condition ensures to get a list of solutions such that we can apply `subs` in them.
Why we need this condition- [reason](https://github.com/sympy/sympy/pull/20845#issuecomment-780678266).
The main issue is with `solve`,
```
In [1]: solve(x**n-1,x)
Out[1]: [1]
```
Here what if `n` was even integer, or maybe any other odd integer except for `n=1`, However it works fine for the particular values when `n` is provided
Kindly share your opinion.
Thanks.

oscarbenjamin replied:
To me that just seems like a bug in `solve`. I would rather not workaround it here but instead just open an issue describing the problem. If `dsolve` doesn't give the complete answer in this case then that can be improved later when solve is improved or if a different solver is used instead.

hyadav2k replied:
Sure, [solve.](https://github.com/sympy/sympy/issues/21014)
Thanks.

oscarbenjamin replied:
I think this case should be removed from `odesimp`. The function should either be fixed in general or left alone. There should not be special checks for particular solvers.

hyadav2k replied:
Indeed, will try to fix for `solve`, probably that will handle the issues and we can omit the `if` condition.
Thanks for replying.
